packages/lit-html/src/lib/directives/class-map.ts

```diff
@@ -0,0 +1,148 @@
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {
+  AttributePart,
+  directive,
+  PropertyPart,
+  Directive,
+  noChange,
+  Part,
+} from '../lit-html.js';
+
+// IE11 doesn't support classList on SVG elements, so we emulate it with a Set
+// TODO (justinfagnani): Can we move this to the IE compat patch?
+class ClassList {
+  element: Element;
+  classes: Set<string> = new Set();
+  changed = false;
+
+  constructor(element: Element) {
+    this.element = element;
+    const classList = (element.getAttribute('class') || '').split(/\s+/);
+    for (const cls of classList) {
+      this.classes.add(cls);
+    }
+  }
+  add(cls: string) {
+    this.classes.add(cls);
+    this.changed = true;
+  }
+
+  remove(cls: string) {
+    this.classes.delete(cls);
+    this.changed = true;
+  }
+
+  commit() {
+    if (this.changed) {
+      let classString = '';
+      this.classes.forEach((cls) => (classString += cls + ' '));
+      this.element.setAttribute('class', classString);
+    }
+  }
+}
+
+export interface ClassInfo {
+  readonly [name: string]: string | boolean | number;
+}
+
+/**
+ * A directive that applies CSS classes. This must be used in the `class`
+ * attribute and must be the only part used in the attribute. It takes each
+ * property in the `classInfo` argument and adds the property name to the
+ * element's `class` if the property value is truthy; if the property value is
+ * falsey, the property name is removed from the element's `class`. For example
+ * `{foo: bar}` applies the class `foo` if the value of `bar` is truthy.
+ * @param classInfo {ClassInfo}
+ */
+export const classMap = directive(
+  class ClassMap extends Directive {
+    /**
+     * Stores the ClassInfo object applied to a given AttributePart.
+     * Used to unset existing values when a new ClassInfo object is applied.
+     */
+    previousClasses?: Set<string>;
+
+    constructor(part: Part) {
+      super();
+      if (
+        !(part instanceof AttributePart) ||
+        part instanceof PropertyPart ||
+        part.name !== 'class' ||
+        (part.__strings !== undefined && part.__strings.length > 2)
+      ) {
+        throw new Error(
+          'The `classMap` directive must be used in the `class` attribute ' +
+            'and must be the only part in the attribute.'
+        );
+      }
+    }
+
+    render(classInfo: ClassInfo) {
+      return Array.from(Object.entries(classInfo))
+        .filter(([_, enabled]) => enabled)
+        .map(([name, _]) => name)
+        .join(' ');
+    }
+
+    update(part: AttributePart, [classInfo]: [ClassInfo]) {
+      // Remember dynamic classes on the first render
+      if (this.previousClasses === undefined) {
+        this.previousClasses = new Set();
+        for (const name in classInfo) {
+          if (classInfo[name]) {
+            this.previousClasses.add(name);
+          }
+        }
+        return this.render(classInfo);
+      }
+
+      const element = part.__element;
```

Comment:
`__element` should be public or passed in separately.

Reply:
We'll make it public